pkgs/operator/state.go

```diff
@@ -220,11 +309,78 @@ func (s *Switch) InitInstance(reqID [24]byte, initMsg *wire.Transport, initiator
 	s.Instances[reqID] = inst
 	s.InstanceInitTime[reqID] = time.Now()
 	s.Mtx.Unlock()
-
 	return resp, nil
+}
 
+func (s *Switch) InitInstanceReshare(reqID [24]byte, reshareMsg *wire.Transport, initiatorSignature []byte) ([]byte, error) {
+	logger := s.Logger.With(zap.String("reqid", hex.EncodeToString(reqID[:])))
+	logger.Info("🚀 Initializing DKG instance")
+	reshare := &wire.Reshare{}
+	if err := reshare.UnmarshalSSZ(reshareMsg.Data); err != nil {
+		return nil, err
+	}
+	// Check that incoming init message signature is valid
+	initiatorPubKey, err := crypto.ParseRSAPubkey(reshare.InitiatorPublicKey)
+	if err != nil {
+		return nil, err
+	}
+	marshalledWireMsg, err := reshareMsg.MarshalSSZ()
+	if err != nil {
+		return nil, err
+	}
+	err = crypto.VerifyRSA(initiatorPubKey, marshalledWireMsg, initiatorSignature)
+	if err != nil {
+		return nil, fmt.Errorf("init message: initiator signature isn't valid: %s", err.Error())
+	}
+	initiatorID := sha256.Sum256(initiatorPubKey.N.Bytes())
+	s.Logger.Info("✅ reshare message signature is successfully verified", zap.String("from initiator", fmt.Sprintf("%x", initiatorID[:])))
+
+	s.Logger.Info("Starting resharing protocol")
+	// try to get old share local owner first
+	var shareFromDB basedb.Obj
+	secret := &kyber_dkg.DistKeyShare{}
+	shareFromDB, ok, err := s.DB.Get([]byte("secret"), reshare.OldID[:])
+	if err != nil {
+		return nil, err
+	}
+	if ok {
+		secret, err = s.GetSecretShare(shareFromDB)
+		if err != nil {
+			return nil, err
+		}
+	}
+	s.Mtx.Lock()
+	l := len(s.Instances)
+	if l >= MaxInstances {
+		cleaned := s.CleanInstances() // not thread safe
+		if l-cleaned >= MaxInstances {
+			s.Mtx.Unlock()
+			return nil, ErrMaxInstances
+		}
+	}
+	_, ok = s.Instances[reqID]
+	if ok {
+		tm := s.InstanceInitTime[reqID]
+		if !time.Now().After(tm.Add(MaxInstanceTime)) {
+			s.Mtx.Unlock()
+			return nil, ErrAlreadyExists
+		}
+		delete(s.Instances, reqID)
+		delete(s.InstanceInitTime, reqID)
```

Comment:
this is potential attack vector I just thought of ( though it was probably my bad ). if a reqid is expired like in this code and someone runs a new instance with this reqid it might rewrite the secret share? are we saving them by id? we should think how to mitigate, might be to not allow reusing a reqid, if its in the list or if its in the db.

Reply:
Yeah looks like its a weak point here. We do store a secret share at DB https://github.com/bloxapp/ssv-dkg/blob/key_resharing/pkgs/dkg/drand.go#L442 using `reqID` as we need it at resharing. 
An dishonest initiator seems can rewrite a share at operators if a new init ceremony with the same ID is being run, but the resulting validator will be different anyway. So looks like the only resharing at risk here.

Reply:
I will add the check if we already have ID at DB - we bail right away.